### ISSUE TITLE
PyTorch Dynamo ONNX Export: Cosmos Video Transformer Model

### ISSUE THREAD

**Issue by asfiyab-nvidia** [pytorch/pytorch#168048]
### 🐛 Describe the bug

This issue extends https://github.com/pytorch/pytorch/issues/163170

cc @yuanyao-nv @titaiwangms @justinchuby @xadupre 

The Transformer model for the Cosmos Video2World pipeline requires the timestep input to be of size `[BS, 1, latent_frames, 1, 1]`. This shape is unsupported by the exporter as shown in the error pasted at the bottom of this post. The export succeeds when timestep is of size `[BS]`.
I used and env with PyTorch nightly and ONNXscript 0.5.4 

Export script is below

```
import torch
from diffusers import CosmosTransformer3DModel

device = "cuda"
dtype = torch.bfloat16
model = CosmosTransformer3DModel.from_pretrained(
    "nvidia/Cosmos-Predict2-2B-Video2World",
    subfolder="transformer",
    use_safetensors=True,
    torch_dtype=dtype,
).to(device)

batch_size = 1
image_height = 704
image_width = 1280
latent_height = image_height // 8
latent_width = image_width // 8
latent_frames = 24
latent_channels = model.config.in_channels - 1
text_maxlen = 512
sample_input = {
    "hidden_states": torch.randn(
        batch_size,
        latent_channels,
        latent_frames,
        latent_height,
        latent_width,
        dtype=dtype,
        device=device,
    ),
    "timestep": torch.tensor([1.0] * batch_size, dtype=dtype, device=device).view(1, 1, 1, 1, 1).expand(
                    batch_size, -1, latent_frames, -1, -1
                ),
    "encoder_hidden_states": torch.randn(
        batch_size,
        text_maxlen,
        model.config["text_embed_dim"],
        dtype=dtype,
        device=device,
    ),
    "padding_mask": torch.ones(
        1, 1, image_height, image_width, dtype=dtype, device=device
    ),
    "fps": torch.tensor([16] * batch_size, dtype=dtype, device=device),
    "condition_mask": torch.randn(
        batch_size,
        1,
        latent_frames,
        latent_height,
        latent_width,
        dtype=dtype,
        device=device,
    ),
}

input_names = [
    "hidden_states",
    "timestep",
    "encoder_hidden_states",
    "padding_mask",
    "fps",
    "condition_mask",
]

output_names = ["latent"]

dynamic_axes = {
    "hidden_states": {0: "B", 2: "latent_frames", 3: "latent_H", 4: "latent_W"},
    "timestep": {0: "B", 2: "latent_frames"},
    "encoder_hidden_states": {0: "B"},
    "padding_mask": {0: "B", 2: "H", 3: "W"},
    "fps": {0: "B"},
    "condition_mask": {0: "B", 2: "latent_frames", 3: "latent_H", 4: "latent_W"},
}


torch.onnx.export(
    model,
    (sample_input,),
    "cosmos_transformer_vid2world/model.onnx",
    export_params=True,
    opset_version=19,
    do_constant_folding=True,
    input_names=input_names,
    output_names=output_names,
    dynamic_shapes=dynamic_axes,
    verbose=False,
    dynamo=True,
)
```

Error stack:
```
## Exception summary

<class 'torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode'>: Could not guard on data-dependent expression Eq(128, u3) (unhinted: Eq(128, u3)).  (Size-like symbols: u3)

consider using data-dependent friendly APIs such as guard_or_false, guard_or_true and statically_known_true.
Caused by: (_ops.py:836 in __call__)
For more information, run with TORCH_LOGS="dynamic"
For extended logs when we create symbols, also add TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="u3"
If you suspect the guard was triggered from C++, add TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
For more debugging help, see https://docs.google.com/document/d/1HSuTTVvYH1pTew89Rtpeu84Ht3nQEFTYhAX3Ypa_xJs/edit?usp=sharing

For C++ stack trace, run with TORCHDYNAMO_EXTENDED_DEBUG_CPP=1

The following call raised this error:
  File "/usr/local/lib/python3.12/dist-packages/diffusers/models/transformers/transformer_cosmos.py", line 203, in __call__
    hidden_states = F.scaled_dot_product_attention(

To fix the error, insert one of the following checks before this call:
  1. torch._check(128 == key.shape[3])
  2. torch._check(128 != key.shape[3])

(These suggested fixes were derived by replacing `u3` with key.shape[3] in Eq(128, u3) and its negation.)

The error above occurred when calling torch.export.export. If you would like to view some more information about this error, and get a list of all other errors that may occur in your export call, you can replace your `export()` call with `draft_export()`.

(Refer to the full stack trace above for more information.)
```

### Versions

PyTorch nightly
ONNXscript 0.5.4

**Comment by titaiwangms** [comment]
Yeah, we have seen this couple times. Basically, we need to add a `torch ._check` to tell shape guards that attn_mask.shape[?] (I forgot which dim represents `seq_len`) is equal to key.shape[3] (right before `hidden_states = F.scaled_dot_product_attention(...`). transformers re-generates their own mask with `masking_utils.py`, and shape guards lose the tracks.right before 

cc @xadupre

**Comment by justinchuby** [comment]
See also https://github.com/huggingface/transformers/pull/41900
